### ISSUE TITLE
Two more detectMimeType overloads 

### ISSUE THREAD

**Issue by melloware** [overview/mime-types#15]
I would like two more overloads of `detectMimeType`

1. Only with a filename and no InputStream currently I do this..

```java
detectMimeType(filename, (InputStream) null);
```

2. Only an InputStream without a filename hint currently I do this.

```java
detectMimeType(null, inputStream);
```

I do this so I can compare a name with its actual content in code.

```java
final String fileNameType = detectTypeByName(filename);
final String contentType = detectTypeByContent(inputStream);

if (!Objects.equals(fileNameType, contentType)) {
    throw new MaliciousFileException(String.format("File '%s' is type '%s' but the content is '%s'.", filename,
                        fileNameType, contentType));
}
```

**Example:** 

> File 'malicious2.gif' is mime type 'image/gif' but the content is 'application/x-ms-dos-executable'.

**Comment by adamhooper** [comment]
I would warn against this pattern: shared-mime-info is not designed for security.

shared-mime-info is designed to select the most appropriate icon and program for a file. If the user changes filename, that means a different icon and program.

For instance: name a Word document ".zip" to open it as an archive.

And if you try and flip it and use magic numbers to detect the file type? Well, without a filename we can't recognize a Word document, because a Word document is just a zip file. It is impossible *in general* to figure out mime type given only contents, and this library doesn't try. Instead, it treat filename as user input, not developer input.

There may be other libraries for security by magic number. Or you can handle this security problem in your code directly. (For instance, if you're building an image-upload tool, verify magic numbers and file extensions.)

Or you can use *this* library with an ambiguous filename like "image" ... but understand what this wins you and what it doesn't! It might guard against people uploading DOS executables instead of PNG/gif/jpeg, but it won't guard against people uploading DOS executables as Word documents. (And anyway, who's going to try executing GIF files?)

For real security, make sure your gif-opening code is secure. You need to do that anyway, regardless of the file's magic number.

I'm tempted to close ... is there something I'm missing?

**Comment by melloware** [comment]
No you have it exactly correct!  I already handle your cases above with an "override" map of NAME/TYPE for example:

`application/vnd.openxmlformats-officedocument.presentationml.presentation = application/zip`

So I know `application/zip` is OK for OpenOffice documents but if it was an `msdos-executable` I would reject it.  GIF might be a bad example but things that can be double clicked on such as PowerPoint should not actually be an EXE.  

The idea i a website or microservice only allowes say "uploading images and OpenOffice docs" i have a way of rejecting all other types and also prevent someone from sticking an .EXE as a .DOC file and have someone click on it.

We have had this

**Comment by melloware** [comment]
I can close this since my code is doing this already so no big deal!

**Comment by melloware** [comment]
I should also mention we are virus scanning our files too but as you know Virus Scanners are only updated AFTER they are aware of malware which is usually too late.  This is just an extra layer of "if you tell me something is a PDF it better be a PDF".

**Comment by adamhooper** [comment]
> This is just an extra layer of "if you tell me something is a PDF it better be a PDF".

Two security tidbits (because it sounds like you're working on something similar to Overview):

1. Even without a virus, a PDF can correctly cost your PDF reader infinite CPU and near-infinite RAM. If your server is reading the PDF, you'll need some sort of sandbox (limited RAM and time -- a custom-made PDF reader, or a normal PDF reader in a virtual machine or container invoked with a timeout).
2. Any PDF reader you can find will check the file's magic number before proceeding.

TL;DR "is this file _really_ a PDF" isn't the question you want to answer. It's, "is this file good or bad?" ... and that's totally different.

**Comment by melloware** [comment]
Yep agreed. Don't disagree with any of that. 

But this magic byte check prevents me from wasting resources scanning it further or uploading to Cloud storage.  If they are uploading a PNG image and if it's not a PNG reject it as fast as possible with magic bytes before wasting any more resources virus scanning it or doing deeper checks on the file.
